src/binder/expression_binder.cpp

```diff
@@ -76,7 +78,15 @@ std::shared_ptr<Expression> ExpressionBinder::implicitCastIfNecessary(
         return expression;
     }
     assert(targetTypeID != VAR_LIST);
-    return implicitCast(expression, DataType(targetTypeID));
+    if (BuiltInVectorOperations::getCastCost(expression->dataType.typeID, targetTypeID) !=
```

Comment:
This is different from what I have in mind.

Both `implicitCastIfNecessary(expression, targetType)` and `implicitCastIfNecessary(expression, targetTypeID)` should
- first check if type is already compatible, in such case we don't need to append a cast function expression on top
- if not, both should enter `implicitCast(expression, targetType)` and append a cast function expression on top of the input expression. If we don't support casting, then a binder exception should be thrown.

The reason why we have two implicitCastIfNecessary is documented in the header. 

Feel free to discuss with me if this comment is not clear.

Reply:
Moved casting logic to implicitCast()